src/modules/commander/Commander.cpp

```diff
@@ -1790,92 +1791,148 @@ void Commander::executeActionRequest(const action_request_s &action_request)
+		const double lat = static_cast<double>(gps_pos.lat) * 1e-7;
+		const double lon = static_cast<double>(gps_pos.lon) * 1e-7;
+		const float alt = static_cast<float>(gps_pos.alt) * 1e-3f;
+		fillGlobalHomePos(home, lat, lon, alt);
+		setHomePosValid();
+		updated = true;
+
+	} else if (_local_position_sub.get().z_global) {
+		// handle special case where we are setting only altitude using local position reference
+		// This might be overwritten by altitude from global or GNSS altitude
+		home.timestamp = hrt_absolute_time();
+		home.alt = _local_position_sub.get().ref_alt;
+		home.valid_alt = true;
 
-			// Ensure that the GPS accuracy is good enough for intializing home
-			if (isGPosGoodForInitializingHomePos(gpos)) {
-				fillGlobalHomePos(home, gpos);
-				setHomePosValid();
-			}
-		}
+		updated = true;
+	}
 
-		_home_pub.update(home);
+	if (updated) {
+		home.timestamp = hrt_absolute_time();
+		home.manual_home = false;
+		updated = _home_pub.update(home);
 	}
 
-	return _status_flags.home_position_valid;
+	return updated;
 }
 
 bool
 Commander::set_in_air_home_position()
 {
-	if (_param_com_home_en.get()
-	    && _status_flags.local_position_valid
+	if (_status_flags.local_position_valid
 	    && _status_flags.global_position_valid) {
 
 		const vehicle_global_position_s &gpos = _global_position_sub.get();
 		home_position_s home{};
+		home = _home_pub.get();
+		home.timestamp = hrt_absolute_time();
+		const vehicle_local_position_s &lpos = _local_position_sub.get();
 
-		// Ensure that the GPS accuracy is good enough for intializing home
-		if (isGPosGoodForInitializingHomePos(gpos)) {
-			home = _home_pub.get();
-			home.timestamp = hrt_absolute_time();
-			const vehicle_local_position_s &lpos = _local_position_sub.get();
-
-			if (_home_pub.get().valid_lpos) {
-				// Back-compute lon, lat and alt of home position given the home
-				// and current positions in local frame
-				MapProjection ref_pos{gpos.lat, gpos.lon};
-				double home_lat;
-				double home_lon;
-				ref_pos.reproject(home.x - lpos.x, home.y - lpos.y, home_lat, home_lon);
-				const float home_alt = gpos.alt + home.z;
-				fillGlobalHomePos(home, home_lat, home_lon, home_alt);
-
-			} else {
-				// Home position in local frame is unknowm, set
-				// home as current position
-				fillLocalHomePos(home, lpos);
-				fillGlobalHomePos(home, gpos);
-			}
+		if (home.valid_lpos) {
+			// Back-compute lon, lat and alt of home position given the home
+			// and current positions in local frame
+			MapProjection ref_pos{gpos.lat, gpos.lon};
+			double home_lat;
+			double home_lon;
+			ref_pos.reproject(home.x - lpos.x, home.y - lpos.y, home_lat, home_lon);
+			const float home_alt = gpos.alt + home.z;
+			fillGlobalHomePos(home, home_lat, home_lon, home_alt);
 
-			setHomePosValid();
-			_home_pub.update(home);
+		} else {
+			// Home position in local frame is unknowm, set
+			// home as current position
+			fillLocalHomePos(home, lpos);
+			fillGlobalHomePos(home, gpos);
```

Comment:
`set_home_position()`?

Reply:
Probably better, yes. I still want to review that `set_in_air_home_position` function to handle the back-computation of the local home position too

Reply:
So better follow-up. It's not incorrect to set stuff separately here.

Reply:
I actually added it here: [24d9254](https://github.com/PX4/PX4-Autopilot/pull/19437/commits/24d9254f11da07c2d8b30440103cc16388c2b2e9)